File: apps/agni-server/src/core/interactions/patrimony/deletePatrimony.ts
import Repository from "@core/adapters/repository";
import { IUsecase } from "../interfaces";
import { UnitOfWorkRepository } from "@core/repositories/unitOfWorkRepository";
import { Patrimony } from "@core/domains/entities/patrimony";
import { ResourceNotFoundError } from "@core/errors/resournceNotFoundError";

export default class DeletePatrimonyUseCase implements IUsecase<string, void> {
    private patrimonyRepo: Repository<Patrimony>
    private unitOfWorkRepo: UnitOfWorkRepository

    constructor(patrimonyRepo: Repository<Patrimony>, unitOfWorkRepo: UnitOfWorkRepository) {
        this.patrimonyRepo = patrimonyRepo
        this.unitOfWorkRepo = unitOfWorkRepo
    }

    async execute(patrimonyId: string): Promise<void> {
        try {
            this.unitOfWorkRepo.start()

            if (!await this.patrimonyRepo.get(patrimonyId))
                throw new ResourceNotFoundError("PATRIMONY_NOT_FOUND")

            await this.patrimonyRepo.delete(patrimonyId)

            this.unitOfWorkRepo.commit()
        } catch(err) {
            this.unitOfWorkRepo.rollback()
            throw err;
        } 
    }
}

File: apps/agni_api/src/main/kotlin/dev/auguste/agni_api/core/usecases/patrimonies/DeletePatrimony.kt
package dev.auguste.agni_api.core.usecases.patrimonies

import dev.auguste.agni_api.core.adapters.repositories.IRepository
import dev.auguste.agni_api.core.adapters.repositories.IUnitOfWork
import dev.auguste.agni_api.core.adapters.dto.QueryFilter
import dev.auguste.agni_api.core.adapters.repositories.query_extend.QueryPatrimonySnapshotExtend
import dev.auguste.agni_api.core.entities.Patrimony
import dev.auguste.agni_api.core.entities.PatrimonySnapshot
import dev.auguste.agni_api.core.usecases.interfaces.IUseCase
import dev.auguste.agni_api.core.usecases.patrimonies.dto.DeletePatrimonyInput
import java.util.UUID

class DeletePatrimony(
    private val patrimonyRepo: IRepository<Patrimony>,
    private val patrimonySnapshotRepo: IRepository<PatrimonySnapshot>,
    private val unitOfWork: IUnitOfWork): IUseCase<DeletePatrimonyInput, Unit> {

    override fun execAsync(input: DeletePatrimonyInput) {
        unitOfWork.execute {
            patrimonyRepo.get(input.patrimonyId) ?: throw Error("Patrimony not found")

            patrimonySnapshotRepo.getAll(query = QueryFilter(0, 0, true), QueryPatrimonySnapshotExtend(setOf(input.patrimonyId)))

            patrimonyRepo.delete(input.patrimonyId)
        }
    }
}